### ISSUE TITLE
API Proposal: ShiftLeft, ShiftRight for Vector<T>

### ISSUE THREAD

**Issue by mjmckp** [dotnet/runtime#20510]
# Updated Proposal

https://github.com/dotnet/runtime/issues/20510#issuecomment-895708361

-------------------------------------
# Original Proposal

This is a proposal of the ideas discussed in issue dotnet/runtime#20147.

There are a few SIMD operations on `System.Numerics.Vector<T>` missing that are required to implement vectorized versions of `log`, `exp`, `sin`, and `cos`.

There is a C implementation of these methods using AVX intrinsics [here](http://software-lisc.fbk.eu/avx_mathfun/), which I have translated into a C# implemention using `Vector<T>` [here](https://github.com/mjmckp/VectorMathFuns).   In the implementation, I have added fake versions of the missing intrinsics.

@mellinoe suggested creating two separate issues for the missing methods (the other issue is dotnet/runtime#20509), so the purpose of this issue is to propose the addition of the following methods:
```
public static Vector<short> ShiftLeft(Vector<short> x, int n)
public static Vector<int> ShiftLeft(Vector<int> x, int n)
public static Vector<long> ShiftLeft(Vector<long> x, int n)
public static Vector<short> ShiftRight(Vector<short> x, int n)
public static Vector<int> ShiftRight(Vector<int> x, int n)
public static Vector<long> ShiftRight(Vector<long> x, int n)
```
which map directly to the following AVX2 intrinsics:
```
_mm256_slli_epi16
_mm256_slli_epi32
_mm256_slli_epi64
_mm256_srli_epi16
_mm256_srli_epi32
_mm256_srli_epi64
```
respectively.

**Comment by tannergooding** [comment]
I'm fine with taking a proposal for `ShiftLeft` and `ShiftRight` additions to `Vector<T>`. Given how we implemented HWIntrinsics and the proposed analyzer: https://github.com/dotnet/runtime/issues/33771, I think the overall implementation will work out.

@mjmckp could you please update the top post to follow our API proposal template? https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.md
I think the currently proposed shape is good, but would recommend also proposing overloads for the unsigned integer types. 

Once that is done, I can mark this as `api-ready-for-review`. However, it is worth noting that full framework will never have this be optimized as it isn't taking runtime changes (this likely would have been the case 3 years ago when it was propoosed as well).

**Comment by wstaelens** [comment]
Any news regarding Vector, << and >> ? 😄  
Could optimize a lot of calculations by using a Vector4 and bitshifting << 8   >> 13 etc...

**Comment by tannergooding** [comment]
Given the OP hasn't responded, if someone wants to spend the time to write up the proposal following our template, then this can move forward. Our template is: https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.md

The interested party can either open a new issue or post it here and I can update the original post.
-- I'm a bit pre-occupied with other work right now, and will get to it after .NET 6 finishes locking down if someone else doesn't do it sooner.

**Comment by tannergooding** [comment]
# Backround and motivation

`Vector<T>` exposes a common abstraction for SIMD acceleration across various hardware platforms and architectures. It exposes many core operations today but notably does not include support for shifting.

# API Proposal
```csharp
namespace System.Numerics
{
    public static class Vector
    {
        public static Vector<byte> ShiftLeft(Vector<byte> x, int n);
        public static Vector<short> ShiftLeft(Vector<short> x, int n);
        public static Vector<int> ShiftLeft(Vector<int> x, int n);
        public static Vector<long> ShiftLeft(Vector<long> x, int n);
        public static Vector<nint> ShiftLeft(Vector<nint> x, int n);
        public static Vector<sbyte> ShiftLeft(Vector<sbyte> x, int n);
        public static Vector<ushort> ShiftLeft(Vector<ushort> x, int n);
        public static Vector<uint> ShiftLeft(Vector<uint> x, int n);
        public static Vector<ulong> ShiftLeft(Vector<ulong> x, int n);

        public static Vector<byte> ShiftRightLogical(Vector<byte> x, int n);
        public static Vector<short> ShiftRightLogical(Vector<short> x, int n);
        public static Vector<int> ShiftRightLogical(Vector<int> x, int n);
        public static Vector<long> ShiftRightLogical(Vector<long> x, int n);
        public static Vector<nint> ShiftRightLogical(Vector<nint> x, int n);
        public static Vector<sbyte> ShiftRightLogical(Vector<sbyte> x, int n);
        public static Vector<ushort> ShiftRightLogical(Vector<ushort> x, int n);
        public static Vector<uint> ShiftRightLogical(Vector<uint> x, int n);
        public static Vector<ulong> ShiftRightLogical(Vector<ulong> x, int n);

        public static Vector<short> ShiftRightArithmetic(Vector<short> x, int n);
        public static Vector<int> ShiftRightArithmetic(Vector<int> x, int n);
        public static Vector<long> ShiftRightArithmetic(Vector<long> x, int n);
        public static Vector<nint> ShiftRightArithmetic(Vector<nint> x, int n);
        public static Vector<sbyte> ShiftRightArithmetic(Vector<sbyte> x, int n);
    }
}
```

# Additional Notes

It would likely be beneficial to expose `shift` operators on `Vector<T>` where right shift defaults to arithmetic for signed integers and logical for unsigned integers (https://github.com/dotnet/csharplang/issues/4682). There is a potential issue with `float` and `double` where they would either need to `throw` or be treated as `int` and `long`, respectively.

The RHS of the shift operator would ideally be `byte` to simplify conversion to the hardware instructions or `T` to simplify usage with non-constants inputs. However, C# currently requires the RHS to be `int` and this is the most natural type for general purpose C# code (https://github.com/dotnet/csharplang/issues/4666).

Variants could be exposed which support "variable shift" (that is where each element is shift by a different amount). However, this isn't available on all hardware and may be a perf pit for some users.

**Comment by rickbrew** [comment]
`Vector<T>` supporting bit shift operations will open up a whole lot of ARM64 performance. Right now I only have the resources for 2 code paths (one SIMD, one scalar), so I usually can't use `Vector<T>` and must use SSE/AVX intrinsics. Without bit-shifts, `Vector<T>` is unfortunately pretty useless for most image processing code.
